File: examples/python/RunAnywhereAI/_env.py
"""Tiny stdlib .env loader + credential-aware initialize for the examples.

Mirrors the Android example's local.properties contract: with both a base URL and
an API key set, initialize in PRODUCTION (org-scoped, authed telemetry); with
neither, initialize in DEVELOPMENT (keyless — telemetry goes to the public org).
Copy .env.example to .env and fill in real values, or export the same names.
"""
from __future__ import annotations

import os
from pathlib import Path

import runanywhere as ra
from runanywhere import Environment


def load_dotenv(path: str | os.PathLike[str] | None = None) -> None:
    """Load KEY=VALUE lines from a .env file into os.environ (never overriding it)."""
    env_path = Path(path) if path else Path(__file__).with_name(".env")
    if not env_path.is_file():
        return
    for raw in env_path.read_text().splitlines():
        line = raw.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, _, value = line.partition("=")
        key, value = key.strip(), value.strip().strip('"').strip("'")
        if key and key not in os.environ:
            os.environ[key] = value


def initialize_from_env() -> None:
    """Initialize the SDK with backend creds from .env / the environment."""
    load_dotenv()
    api_key = os.environ.get("RUNANYWHERE_API_KEY", "").strip() or None
    base_url = os.environ.get("RUNANYWHERE_BASE_URL", "").strip() or None
    # Keyed → production (authed telemetry); keyless → development (public org).
    environment = (
        Environment.PRODUCTION if (api_key and base_url) else Environment.DEVELOPMENT
    )
    ra.initialize(api_key=api_key, base_url=base_url, environment=environment)

File: examples/electron/RunAnywhereAI/src/main/env.ts
/**
 * Control-plane credentials, read from a gitignored `.env` (or the environment).
 *
 * Mirrors the Android example's `local.properties` contract: with both a base URL
 * and an API key set, the SDK initializes in PRODUCTION (org-scoped, authed
 * telemetry); with neither, DEVELOPMENT (keyless).
 *
 * Read in main because the sandboxed renderer has no filesystem access.
 */
import fs from 'node:fs';
import path from 'node:path';

import type { BackendConfig } from '../shared/ipc-contract';

import { APP_ROOT } from './paths';

/** Parse a minimal .env: `KEY=value`, `#` comments, optional surrounding quotes. */
function readDotEnv(file: string): Record<string, string> {
  const out: Record<string, string> = {};
  let raw: string;
  try {
    raw = fs.readFileSync(file, 'utf8');
  } catch {
    return out; // no .env is fine — fall back to process.env / keyless development
  }
  for (const line of raw.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (trimmed === '' || trimmed.startsWith('#') || !trimmed.includes('=')) continue;
    const eq = trimmed.indexOf('=');
    const key = trimmed.slice(0, eq).trim();
    const value = trimmed
      .slice(eq + 1)
      .trim()
      .replace(/^['"]|['"]$/g, '');
    if (key !== '') out[key] = value;
  }
  return out;
}

export function backendConfig(): BackendConfig {
  const fromFile = readDotEnv(path.join(APP_ROOT, '.env'));
  // A real environment variable wins over the file, so a CI or shell override
  // does not need the file edited.
  const read = (key: string): string => (process.env[key] ?? fromFile[key] ?? '').trim();

  const apiKey = read('RUNANYWHERE_API_KEY');
  const baseUrl = read('RUNANYWHERE_BASE_URL');

  return {
    apiKey,
    baseUrl,
    environment: apiKey !== '' && baseUrl !== '' ? 'production' : 'development',
  };
}
